scripts/build.py

```diff
@@ -715,63 +858,72 @@ def generate_build_directory():
-        # If there is no staging dir, perform build process on all files.
-        print 'Creating new %s folder' % TEMPLATES_STAGING_DIR
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes)
-    else:
-        # If there is an existing staging dir, rebuild all HTML files.
-        print 'Staging directory exists, re-building all HTML files'
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes,
-            file_formats=('.html',))
-        recently_changed_filenames = get_recently_changed_filenames(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_OUT_DIR)
-        if recently_changed_filenames:
-            # Only re-build files that have changed since last build.
-            print ("Re-building recently changed files at %s"
-                   % TEMPLATES_DEV_DIR_CORE)
-            rebuild_new_files(
-                TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR,
-                recently_changed_filenames, hashes)
-            # Clean up files in staging dir that have been removed in DEV dir.
-            remove_deleted_files(TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR)
-        else:
-            print 'No changes detected. Using previously built files.'
-    # Copy built files from staging dir to build dir.
-    copy_files_source_to_target(
-        TEMPLATES_STAGING_DIR, TEMPLATES_OUT_DIR, hashes)
-
+    build_directory(TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS, hashes, build_tasks)
+    _execute_tasks(build_tasks)
+
+    # Copy all files from staging directory to production directory (/build).
+    COPY_INPUT_DIRS = [
+        ASSETS_DEV_DIR, EXTENSIONS_DIRNAMES_TO_DIRPATHS['staging_dir'],
+        TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['staging_dir'],
+        THIRD_PARTY_GENERATED_STAGING_DIR]
+    COPY_OUTPUT_DIRS = [
+        ASSETS_OUT_DIR, EXTENSIONS_DIRNAMES_TO_DIRPATHS['out_dir'],
+        TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['out_dir'],
+        THIRD_PARTY_GENERATED_OUT_DIR]
+    assert len(COPY_INPUT_DIRS) == len(COPY_OUTPUT_DIRS)
+    for i in xrange(len(COPY_INPUT_DIRS)):
+        generate_copy_tasks_to_copy_from_source_to_target(
+            COPY_INPUT_DIRS[i], COPY_OUTPUT_DIRS[i], hashes, copy_tasks)
+    _execute_tasks(copy_tasks)
+
+    for i in xrange(len(COPY_INPUT_DIRS)):
+        # Make sure that all files in source directory and staging directory are
+        # accounted for.
+        _compare_file_count(COPY_INPUT_DIRS[i], COPY_OUTPUT_DIRS[i])
+    # Make sure that hashed file name matches with current hash dict.
+    for built_dir in COPY_OUTPUT_DIRS:
+        for root, _, files in os.walk(built_dir):
+            for filename in files:
+                parent_dir = os.path.basename(root)
+                converted_filepath = os.path.join(
+                    THIRD_PARTY_GENERATED_STAGING_DIR, parent_dir, filename)
+                if not hash_should_be_inserted(converted_filepath):
+                    # These filenames do not contain hashes.
```

Comment:
do not contain hashes --> should not be hashed (I think?)

Reply:
Changed comment to reflect if block logic (should not be hashed)